src/Compilers/Core/Portable/MetadataReader/PEModule.cs

```diff
@@ -1623,26 +1661,56 @@ private bool HasStringAndIntValuedAttribute(EntityHandle token, AttributeDescrip
             }
         }
 
-        private static bool CrackObsoleteAttributeData(out ObsoleteAttributeData value, ref BlobReader sig)
+#nullable enable
+        /// <summary>
+        /// Gets the well-known optional named properties on ObsoleteAttribute, if present.
+        /// Both 'diagnosticId' and 'urlFormat' may be present, or only one, or neither.
+        /// </summary>
+        /// <remarks>
+        /// Failure to find any of these properties does not imply failure to decode the ObsoleteAttribute,
+        /// so we don't return a value indicating success or failure.
+        /// </remarks>
+        private static (string? diagnosticId, string? urlFormat) CrackObsoleteProperties(ref BlobReader sig, IAttributeNamedArgumentDecoder decoder)
         {
-            string message;
-            if (CrackStringInAttributeValue(out message, ref sig) && sig.RemainingBytes >= 1)
+            string? diagnosticId = null;
+            string? urlFormat = null;
+
+            try
             {
-                bool isError = sig.ReadBoolean();
-                value = new ObsoleteAttributeData(ObsoleteAttributeKind.Obsolete, message, isError);
-                return true;
+                // See CIL spec section II.23.3 Custom attributes
+                //
+                // Next is a description of the optional “named” fields and properties.
+                // This starts with NumNamed– an unsigned int16 giving the number of “named” properties or fields that follow.
+                var numNamed = sig.ReadUInt16();
+                for (int i = 0; i < numNamed && (diagnosticId is null || urlFormat is null); i++)
+                {
+                    var ((name, value), isProperty, typeCode) = decoder.DecodeCustomAttributeNamedArgumentOrThrow(ref sig);
+                    if (typeCode == SerializationTypeCode.String && isProperty && value.ValueInternal is string stringValue)
+                    {
+                        if (diagnosticId is null && name == "DiagnosticId")
+                        {
+                            diagnosticId = stringValue;
+                        }
+                        else if (urlFormat is null && name == "UrlFormat")
+                        {
+                            urlFormat = stringValue;
+                        }
+                    }
+                }
             }
+            catch (BadImageFormatException) { }
+            catch (UnsupportedSignatureContent) { }
```

Comment:
In the case of either exception should we clear `diagnosticId` and `urlFormat`? Not clearing means that in a corrupted image we could end up reading one tag but not the other. Instinct says that in a corrupted image we should read nothing.

Reply:
We could reject all the named properties on the attribute, on the other hand we could reject the attribute entirely, and not issue any Obsolete diagnostics on it for instance. Does that seem like the right thing?

Reply:
> We could reject all the named properties on the attribute, on the other hand we could reject the attribute entirely, and not issue any Obsolete diagnostics on it for instance. Does that seem like the right thing?

I think it is reasonable to use values that we could decode even though we couldn't decode some values after that, due to, for example, a missing assembly reference needed for that. Silently ignoring all the named args, or ignoring the entire attribute, wouldn't add much value for a consumer, IMHO.   

---
In reply to: [399574171](https://github.com/dotnet/roslyn/pull/42518#discussion_r399574171) [](ancestors = 399574171)